File: JVM/jnetproto-java/src/main/java/jnetproto/java/BinaryWriter.java
package jnetproto.java;

import jnetproto.java.compat.BitConverter;
import jnetproto.java.compat.Reflect;
import org.javatuples.Tuple;

import java.io.IOException;
import java.io.OutputStream;
import java.lang.reflect.Array;
import java.math.BigDecimal;
import java.nio.charset.Charset;
import java.time.Duration;
import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.UUID;

public class BinaryWriter implements IDataWriter {
    private final Charset _enc;
    private final DateTimeFormatter _fmt;
    private final OutputStream _stream;

    public BinaryWriter(OutputStream stream) {
        _enc = Charset.forName("UTF8");
        _fmt = DateTimeFormatter.ofPattern("SSSSSSS");
        _stream = stream;
    }

    @Override
    public void writeBool(boolean value) throws IOException {
        writeI8((byte) (value ? 1 : 0));
    }

    @Override
    public void writeI8(byte value) throws IOException {
        _stream.write(new byte[]{value});
    }

    @Override
    public void writeI16(short value) throws IOException {
        _stream.write(BitConverter.getBytes(value));
    }

    @Override
    public void writeI32(int value) throws IOException {
        _stream.write(BitConverter.getBytes(value));
    }

    @Override
    public void writeI64(long value) throws IOException {
        _stream.write(BitConverter.getBytes(value));
    }

    @Override
    public void writeF32(float value) throws IOException {
        _stream.write(BitConverter.getBytes(value));
    }

    @Override
    public void writeF64(double value) throws IOException {
        _stream.write(BitConverter.getBytes(value));
    }

    @Override
    public void writeF128(BigDecimal value) throws IOException
    {
        var raw = value.toString();
        writeUtf8(raw);
    }

    @Override
    public void writeChar(char value) throws IOException
    {
        writeI16((short) value);
    }

    @Override
    public void writeUtf8(String value) throws IOException
    {
        var bytes = value.getBytes(_enc);
        _stream.write((byte)bytes.length);
        _stream.write(bytes);
    }

    @Override
    public void writeDuration(Duration value) throws IOException {
        var raw = (double) value.toMillis();
        writeF64(raw);
    }

    @Override
    public void writeTimestamp(LocalDateTime value) throws IOException {
        var date = value.atZone(ZoneOffset.UTC);
        writeI64(date.toEpochSecond());
        writeI32(Integer.parseInt(date.format(_fmt)));
    }

    @Override
    public void writeGuid(UUID value) throws IOException {
        _stream.write(BitConverter.getBytes(value));
    }

    @Override
    public void writeArray(Object value) throws IOException {
        var rank = Reflect.getRank(value);
        for (var dim = 0; dim < rank; dim++)
            writeI32(Array.getLength(value));
        for (int i = 0; i < Array.getLength(value); i++) {
            var item = Array.get(value, i);
            writeObject(item, true);
        }
    }

    @Override
    public void writeMap(Map value) throws IOException {
        writeI32(value.size());
        for (var item : value.entrySet()) {
            var entry = (Map.Entry) item;
            writeObject(entry.getKey(), true);
            writeObject(entry.getValue(), true);
        }
    }

    @Override
    public void writeTuple(Tuple value) throws IOException {
        var array = value.toArray();
        writeI8((byte)array.length);
        for (var i = 0; i < array.length; i++)
        {
            writeObject(array[i], false);
        }
    }

    @Override
    public void writeSet(Set value) throws IOException {

    }

    @Override
    public void writeList(List value) throws IOException {

    }

    @Override
    public void writeBag(Object[] value) throws IOException {

    }

    @Override
    public void writeBinary(byte[] value) throws IOException {
        writeI32(value.length);
        _stream.write(value);
    }

    @Override
    public void writeObject(Object value) throws IOException {
        writeObject(value, false);
    }

    private void writeObject(Object value, boolean skipHeader) throws IOException {
        var kind = DataTypes.getKind(value);
        if (!skipHeader)
        {
            _stream.write(Reflect.getByte(kind));
            if (kind instanceof DataTypes.ArrayDt adt)
            {
                _stream.write(Reflect.getByte(adt.Item()));
                _stream.write((byte)adt.Rank());
            }
            else if (kind instanceof DataTypes.MapDt mdt)
            {
                _stream.write(Reflect.getByte(mdt.Key()));
                _stream.write(Reflect.getByte(mdt.Val()));
            }
        }
        switch (kind.Kind())
        {
            case Bool: writeBool((boolean)value); break;
            case I8: writeI8((byte)value); break;
            case I16: writeI16((short)value); break;
            case I32: writeI32((int)value); break;
            case I64: writeI64((long)value); break;
            case F32: writeF32((float)value); break;
            case F64: writeF64((double)value); break;
            case F128: writeF128((BigDecimal)value); break;
            case Char: writeChar((char)value); break;
            case UTF8: writeUtf8((String)value); break;
            case Duration: writeDuration((Duration) value); break;
            case Timestamp: writeTimestamp((LocalDateTime) value); break;
            case Guid: writeGuid((UUID) value); break;
            case Array: writeArray(value); break;
            case Map: writeMap((Map) value); break;
            case Tuple: writeTuple((Tuple) value); break;
            default: throw new IllegalArgumentException(kind.toString());
        }
    }

    @Override
    public void close() throws Exception {
        _stream.close();
    }
}

File: NET/JNetProto.Sharp/Core/BinaryWriter.cs
using System;
using System.Collections;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Text;
using JNetProto.Sharp.API;

namespace JNetProto.Sharp.Core
{
    public class BinaryWriter : IDataWriter
    {
        private readonly CultureInfo _cult;
        private readonly Encoding _enc;
        private readonly Stream _stream;

        public BinaryWriter(Stream stream)
        {
            _cult = CultureInfo.InvariantCulture;
            _enc = Encoding.UTF8;
            _stream = stream;
        }

        public void WriteBool(bool value)
        {
            WriteI8((byte)(value ? 1 : 0));
        }

        public void WriteI8(byte value)
        {
            _stream.Write(new[] { value });
        }

        public void WriteI16(short value)
        {
            _stream.Write(BitConverter.GetBytes(value));
        }

        public void WriteI32(int value)
        {
            _stream.Write(BitConverter.GetBytes(value));
        }

        public void WriteI64(long value)
        {
            _stream.Write(BitConverter.GetBytes(value));
        }

        public void WriteF32(float value)
        {
            _stream.Write(BitConverter.GetBytes(value));
        }

        public void WriteF64(double value)
        {
            _stream.Write(BitConverter.GetBytes(value));
        }

        public void WriteF128(decimal value)
        {
            var raw = value.ToString(_cult);
            WriteUtf8(raw, wide: false);
        }

        public void WriteChar(char value)
        {
            WriteI16((short)value);
        }

        public void WriteUtf8(string value)
        {
            WriteUtf8(value, wide: true);
        }

        private void WriteUtf8(string value, bool wide)
        {
            var bytes = _enc.GetBytes(value);
            if (wide)
                WriteI16((short)bytes.Length);
            else
                WriteI8((byte)bytes.Length);
            _stream.Write(bytes);
        }

        public void WriteDuration(TimeSpan value)
        {
            WriteF64(value.TotalMilliseconds);
        }

        public void WriteTimestamp(DateTime value)
        {
            var date = new DateTimeOffset(value.ToUniversalTime());
            WriteI64(date.ToUnixTimeSeconds());
            WriteI32(int.Parse(date.ToString("fffffff")));
        }

        public void WriteGuid(Guid value)
        {
            _stream.Write(value.ToByteArray());
        }

        public void WriteArray(Array value)
        {
            for (var dim = 0; dim < value.Rank; dim++)
                WriteI32(value.GetLength(dim));
            foreach (var item in value)
                WriteObject(item, true);
        }

        public void WriteMap(IDictionary value)
        {
            WriteI32(value.Count);
            foreach (DictionaryEntry entry in value)
            {
                WriteObject(entry.Key, true);
                WriteObject(entry.Value, true);
            }
        }

        public void WriteSet(IEnumerable value)
        {
            WriteEnumerable(value, true);
        }

        public void WriteList(IList value)
        {
            WriteEnumerable(value, true);
        }

        public void WriteBag(object[] value)
        {
            WriteI8((byte)value.Length);
            foreach (var item in value)
                WriteObject(item, false);
        }

        public void WriteBinary(byte[] value)
        {
            WriteI32(value.Length);
            _stream.Write(value);
        }

        private void WriteEnumerable(IEnumerable raw, bool skipHeader)
        {
            int count;
            IEnumerable values;
            if (raw is ICollection coll)
            {
                count = coll.Count;
                values = raw;
            }
            else
            {
                var t = raw.GetType();
                if (t.GetProperty(nameof(ICollection.Count))?.GetValue(raw) is int c)
                {
                    count = c;
                    values = raw;
                }
                else
                {
                    var array = raw.OfType<object>().ToArray();
                    count = array.Length;
                    values = array;
                }
            }
            WriteI32(count);
            foreach (var entry in values)
                WriteObject(entry, skipHeader);
        }

        public void WriteTuple(ITuple value)
        {
            WriteI8((byte)value.Length);
            for (var i = 0; i < value.Length; i++)
            {
                WriteObject(value[i], false);
            }
        }
        
        public void WriteObject(object value)
        {
            WriteObject(value, false);
        }

        private void WriteObject(object value, bool skipHeader)
        {
            var kind = DataTypes.GetKind(value);
            if (!skipHeader)
            {
                _stream.WriteByte((byte)kind.Kind);
                if (kind is DataTypes.ArrayDt adt)
                {
                    _stream.WriteByte((byte)adt.Item.Kind);
                    _stream.WriteByte((byte)adt.Rank);
                }
                else if (kind is DataTypes.MapDt mdt)
                {
                    _stream.WriteByte((byte)mdt.Key.Kind);
                    _stream.WriteByte((byte)mdt.Val.Kind);
                }
                else if (kind is DataTypes.ListDt ldt)
                {
                    _stream.WriteByte((byte)ldt.Item.Kind);
                }
            }
            switch (kind.Kind)
            {
                case DataType.Bool: WriteBool((bool)value); break;
                case DataType.I8: WriteI8((byte)value); break;
                case DataType.I16: WriteI16((short)value); break;
                case DataType.I32: WriteI32((int)value); break;
                case DataType.I64: WriteI64((long)value); break;
                case DataType.F32: WriteF32((float)value); break;
                case DataType.F64: WriteF64((double)value); break;
                case DataType.F128: WriteF128((decimal)value); break;
                case DataType.Char: WriteChar((char)value); break;
                case DataType.UTF8: WriteUtf8((string)value); break;
                case DataType.Duration: WriteDuration((TimeSpan)value); break;
                case DataType.Timestamp: WriteTimestamp((DateTime)value); break;
                case DataType.Guid: WriteGuid((Guid)value); break;
                case DataType.Array: WriteArray((Array)value); break;
                case DataType.Map: WriteMap((IDictionary)value); break;
                case DataType.Tuple: WriteTuple((ITuple)value); break;
                case DataType.Set: WriteSet((IEnumerable)value); break;
                case DataType.List: WriteList((IList)value); break;
                case DataType.Bag: WriteBag((object[])value); break;
                case DataType.Binary: WriteBinary((byte[])value); break;
                default: throw new ArgumentException(kind.ToString());
            }
        }

        public void Flush()
        {
            _stream.Flush();
        }

        public void Dispose()
        {
            _stream?.Dispose();
        }
    }
}